### PR TITLE
Adding LuMamba to BioFoundation

### PR DESCRIPTION
**Pull request by danaebroustail**
**Overview:**
Added dataset classes, preprocessing code, pretraining and finetuning configs and tasks (LeJEPA pretraining), docs, logo, and model files for **LuMamba**.

**Changes in existing parts of the repo:**
Overlapping changes that might need an extra look:
- **LUNA's `multiloader_data_module.py`**: added option to track t-SNE of TUAB and TUAR during pre-training.
     - Should not affect regular LUNA pre-training, and can be commented out of LuMamba pretrain config. 
-  **Config `defaults.yaml`**: added a hydra dir option, to change hydra output logs directory.
- **LUNA's `finetune_task_LUNA.py`**: added integration for LuMamba, is consistent with LUNA. 
   - Included bug fix for "model." prefixes for LUNA/LuMamba checkpoint loading (.ckpt and .safetensors files). Consistent with LUNA safetensors behavior.

### CODE REVIEW

File: datasets/mumtaz_dataset.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+def get_channel_locations(channel_names):
+    montage = mne.channels.make_standard_montage('GSN-HydroCel-129')
+    ch_pos = montage.get_positions()['ch_pos']
+    locations = {ch: ch_pos[ch] for ch in channel_names if ch in ch_pos}
+    # return as ordered list of locations
+    ordered_locations = [locations[ch] for ch in channel_names if ch in locations]
+    return np.stack(ordered_locations, axis=0)
+
+class MODMA_Dataset(torch.utils.data.Dataset):
+    def __init__(self, hdf5_file, num_channels=128, finetune=True, channel_subset=False):
```
[the remark this answers is not in the document]
Comment (Thoriri) [reply datasets/mumtaz_dataset.py]: @danaebroustail agree? Let's change to modma?

[the remark this answers is not in the document]
Comment (danaebroustail) [reply datasets/mumtaz_dataset.py]: Fixed and replaced MODMA dataset class by Mumtaz dataset wrapper.

File: tasks/pretrain_task_LuMamba.py [hunk 2]
```diff
@@ -0,0 +1,845 @@
...
+import torch_optimizer as torch_optim
+import torch.nn.functional as F
+import matplotlib.pyplot as plt
+import numpy as np
+import wandb
+from einops import rearrange, repeat
+from criterion.query_specialization_criterion import QuerySpecializationCriterion
+from sklearn.manifold import TSNE
+from models.modules.channel_embeddings import CHANNEL_IDX_TO_NAMES
+import psutil, os
```
[the remark this answers is not in the document]
Comment (danaebroustail) [reply tasks/pretrain_task_LuMamba.py:33]: Added wandb as an explicit requirement.

File: tasks/pretrain_task_LuMamba.py [hunk 3]
```diff
@@ -0,0 +1,845 @@
...
+        D = a_emb.size(-1)
+        K = D
+
+        # compute mean along L dimension
+        a_emb = a_emb.mean(dim=1)  # (BV, D)
+        g_emb = g_emb.mean(dim=1)  # (BV, D)
+
+        # rearrange to (V, B, D)
+        a_emb = a_emb.view(V_local, bs, -1)
+        g_emb = g_emb.view(V_global, bs, -1)
```
[the remark this answers is not in the document]
Comment (danaebroustail) [reply tasks/pretrain_task_LuMamba.py]: Fixed by explicitly rearranging dimensions with einops.

File: models/LuMamba.py [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+        self.activation1 = nn.GELU()
+
+    def forward(self, x):
+        # Input x shape: (B, T=grid_size[1], D=embed_dim * grid_size[0])
+        x=x.permute(0,2,1) 
+        x = x.mean(dim=-1)  # Temporal Pooling -> output: (B, embed_dim * grid_size[0])
+
+        # First linear layer: embed_size * grid_size[0] -> num_classes
+        x = self.fc1(x)
+        x = self.activation1(x)
```
[the remark this answers is not in the document]
Comment (danaebroustail) [reply models/LuMamba.py:43]: Fixed by adding return line.

File: make_datasets/make_apava_dataset.py [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+#* Unless required by applicable law or agreed to in writing, software        *
+#* distributed under the License is distributed on an "AS IS" BASIS,          *
+#* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   *
+#* See the License for the specific language governing permissions and        *
+#* limitations under the License.                                             *
+#*                                                                            *
+#* Author:  Danaé Broustail                                                   *
+#* Author:  Thorir Mar Ingolfsson                                             *
+#*----------------------------------------------------------------------------*
+from util.preprocess_utils import segment_raw, write_segments_to_h5_append
```
[the remark this answers is not in the document]
Comment (danaebroustail) [reply make_datasets/make_apava_dataset.py:20]: Ignored for now since the script is executed on the command line and behaviour is as expected.

File: models/LuMamba.py [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+#*                                                                            *
+#* Author:  Danaé Broustail                                                        *
+#* Author:  Thorir Mar Ingolfsson                                             *
+#*----------------------------------------------------------------------------*
+from models.FEMBA import MambaWrapper, PatchEmbed, MambaClassifier, Decoder
+from models.LUNA import *
+import torch.nn as nn
+import torch
+from typing import Tuple
+from mamba_ssm import Mamba
```
[the remark this answers is not in the document]
Comment (danaebroustail) [reply models/LuMamba.py]: Removed import line.

File: tasks/pretrain_task_LuMamba.py [hunk 7]
```diff
@@ -0,0 +1,845 @@
...
+        # Randomly select which rectangles to mask
+        selected_indices = torch.randperm(num_rectangles)[:num_to_mask]
+
+        mask = torch.zeros(batch_size, C, T, dtype=torch.bool).to(self.device)
+
+        # Set mask to True in the selected regions
+        for idx in selected_indices:
+            r, c = rectangles[idx]
+            mask[:, r:r + patch_H, c:c + patch_W] = True  
+
```
[the remark this answers is not in the document]
Comment (danaebroustail) [reply tasks/pretrain_task_LuMamba.py:119]: Ignored, since functionality is preserved. In this way LuMamba is also kept consistent with LUNA code.

File: tasks/pretrain_task_LuMamba.py [hunk 8]
```diff
@@ -0,0 +1,845 @@
...
+        self.criterion = hydra.utils.instantiate(self.hparams.criterion)
+        self.query_specialization_criterion = QuerySpecializationCriterion(**self.hparams.query_specialization_criterion)
+        self.patch_size = self.hparams.masking.patch_size
+        self.masking_ratio = self.hparams.masking.masking_ratio
+        self.unmasked_loss_coeff = self.hparams.masking.unmasked_loss_coeff
+        # Enable normalization if specified in parameters
+        if self.hparams.input_normalization is not None and self.hparams.input_normalization.normalize:
+            self.normalize = True
+            self.normalize_fct = ChannelWiseNormalize()
+
```
[the remark this answers is not in the document]
Comment (danaebroustail) [reply tasks/pretrain_task_LuMamba.py:74]: Added self.normalize = False in __init__ for both LUNA and LuMamba.

File: make_datasets/make_tdbrain_dataset.py [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+#* Unless required by applicable law or agreed to in writing, software        *
+#* distributed under the License is distributed on an "AS IS" BASIS,          *
+#* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   *
+#* See the License for the specific language governing permissions and        *
+#* limitations under the License.                                             *
+#*                                                                            *
+#* Author:  Danaé Broustail                                                   *
+#* Author:  Thorir Mar Ingolfsson                                             *
+#*----------------------------------------------------------------------------*
+from util.preprocess_utils import segment_raw, write_segments_to_h5_append
```
[the remark this answers is not in the document]
Comment (danaebroustail) [reply make_datasets/make_tdbrain_dataset.py:20]: Unchanged (as mentionned for other make_dataset comment)

File: tasks/pretrain_task_LuMamba.py [hunk 10]
```diff
@@ -0,0 +1,845 @@
...
+        T = torch.trapz(err, t, dim=1) * total_N
+
+        T_statistic = T.mean()
+
+        return T_statistic
+
+    def log_lejepa_views(self, x, channel_names, global_views, all_views, start_times_global, start_times_local):
+
+        # iterate over channels
+        sample_idx = 9
```
[the remark this answers is not in the document]
Comment (danaebroustail) [reply tasks/pretrain_task_LuMamba.py:274]: Non applicable. "channel_names" is provided in the batch.

File: tasks/finetune_regression_task_LuMamba.py [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+        self.test_r2(y_pred, y)
+        self.test_rmse(y_pred, y)
+        self.test_pearson(y_pred, y)
+
+        # Log metrics
+        self.log("test_loss", loss, on_step=False, on_epoch=True, prog_bar=True, sync_dist=True)
+        self.log("test_r2", self.test_r2, on_step=False, on_epoch=True, prog_bar=True, sync_dist=True)
+        self.log("test_rmse", self.test_rmse, on_step=False, on_epoch=True, prog_bar=True, sync_dist=True)
+        self.log("test_pearson", torch.mean(self.test_pearson.compute()), on_step=False, on_epoch=True, prog_bar=True, sync_dist=True)
+
```
[the remark this answers is not in the document]
Comment (danaebroustail) [reply tasks/finetune_regression_task_LuMamba.py:217]: Kept as is, it is a workaround for logging the Pearson coefficient metric across a multi-output vector.

File: tasks/pretrain_task_LuMamba.py [hunk 12]
```diff
@@ -0,0 +1,845 @@
...
+        if hasattr(self.logger.experiment, 'log'):  # wandb
+            self.logger.experiment.log({
+                f"LeJEPA_2D_batch_0": wandb.Image(fig),
+                "epoch": self.current_epoch
+            })
+        elif hasattr(self.logger.experiment, 'add_figure'):  # TensorBoard
+            tag = f"LeJEPA_2D_batch_0"
+            self.logger.experiment.add_figure(tag, fig,
+                                              global_step=self.current_epoch)
+
```
[the remark this answers is not in the document]
Comment (danaebroustail) [reply tasks/pretrain_task_LuMamba.py:190]: Kept as is.

File: tasks/finetune_regression_task_LuMamba.py [hunk 13]
```diff
@@ -0,0 +1,271 @@
...
+        self.val_r2(y_pred, y)
+        self.val_rmse(y_pred, y)
+        self.val_pearson(y_pred, y)
+
+        # Log metrics
+        self.log("val_loss", loss, on_step=False, on_epoch=True, prog_bar=True, sync_dist=True)
+        self.log("val_r2", self.val_r2, on_step=False, on_epoch=True, prog_bar=True, sync_dist=True)
+        self.log("val_rmse", self.val_rmse, on_step=False, on_epoch=True, prog_bar=True, sync_dist=True)
+        self.log("val_pearson", torch.mean(self.val_pearson.compute()), on_step=False, on_epoch=True, prog_bar=True, sync_dist=True)
+
```
[the remark this answers is not in the document]
Comment (danaebroustail) [reply tasks/finetune_regression_task_LuMamba.py:193]: Kept as is, it is a workaround for logging the Pearson coefficient metric across a multi-output vector.

File: tasks/finetune_regression_task_LuMamba.py [hunk 14]
```diff
@@ -0,0 +1,271 @@
...
+        self.train_r2(y_pred, y)
+        self.train_rmse(y_pred, y)
+        self.train_pearson(y_pred, y)
+
+        # Log metrics per step (Lightning handles aggregation / syncing)
+        self.log("train_loss", loss, on_step=True, on_epoch=True, prog_bar=True, logger=True, sync_dist=True)
+        self.log("train_r2", self.train_r2, on_step=False, on_epoch=True, prog_bar=True, sync_dist=True)
+        self.log("train_rmse", self.train_rmse, on_step=False, on_epoch=True, prog_bar=True, sync_dist=True)
+        self.log("train_pearson", torch.mean(self.train_pearson.compute()), on_step=False, on_epoch=True, prog_bar=True, sync_dist=True)
+
```
[the remark this answers is not in the document]
Comment (danaebroustail) [reply tasks/finetune_regression_task_LuMamba.py:169]: Ignored, same comment as for train logging. Kept as is, it is a workaround for logging the Pearson coefficient metric across a multi-output vector.
